asteroid/engine/gan_system.py

```diff
@@ -0,0 +1,161 @@
+from collections import OrderedDict
+import soundfile as sf
+import pytorch_lightning as pl
+import torch
+import os
+
+
+class GanSystem(pl.LightningModule):
+
+    def __init__(self, discriminator, generator, opt_d, opt_g,
+                 scheduler_d, scheduler_g, discriminator_loss, generator_loss,
+                 validation_loss, train_loader, val_loader=None, conf=None):
+
+        super(GanSystem, self).__init__()
+        self.discriminator = discriminator
+        self.generator = generator
+        self.opt_d = opt_d
+        self.opt_g = opt_g
+        self.d_loss = discriminator_loss
+        self.g_loss = generator_loss
+        self.validation_loss = validation_loss
+        self.train_loader = train_loader
+        self.val_loader = val_loader
+        self.scheduler_d = scheduler_d
+        self.scheduler_g = scheduler_g
+        self.conf = conf
+
+    def forward(self, z):
+        return self.generator(z)
+
+    def routine_step(self, batch, optimizer_idx):
+
+        # Get data from data_loader
+        inputs, targets = batch
+        # Forward inputs
+        estimates = self(inputs)
+        # The discriminator is basically a binary classifier
+        true_labels = True
+        false_labels = False
+
+        # Train discriminator
+        if optimizer_idx == 0:
+            # Compute D loss for targets (labels = 1)
+            est_true_labels = self.discriminator(targets, inputs, targets)
+            true_loss = self.d_loss(inputs, targets, estimates,
+                                    est_true_labels, true_labels)
+            # Compute D loss for estimates (labels = 0)
+            est_false_labels = self.discriminator(estimates.detach(), inputs,
+                                                  targets)
+            fake_loss = self.d_loss(inputs, targets, estimates,
+                                    est_false_labels, false_labels)
```

Comment:
I don't like that many positional arguments, it's very error-prone. Two or three would be a maximum, five is too much for me.

Reply:
I understand your point but I don't know how I can do a generic call to d_loss with less positional args it's the same problem as my previous comment